File: Editorial/0050_Pow_x_n/Go/1_Binary_Exponentiation.go
func binaryExp(x float64, n int64) float64 {
    // Base case, to stop recursive calls.
    if n == 0 {
        return 1
    }
    // Handle case where, n < 0.
    if n < 0 {
        return 1 / binaryExp(x, -1*n)
    }
    // Perform Binary Exponentiation.
    // If 'n' is odd we perform Binary Exponentiation on 'n - 1' and multiply result with 'x'.
    if n%2 == 1 {
        return x * binaryExp(x*x, (n-1)/2)
    }
    // Otherwise we calculate result by performing Binary Exponentiation on 'n'.
    return binaryExp(x*x, n/2)
}

func myPow(x float64, n int) float64 {
    return binaryExp(x, int64(n))
}

File: Editorial/0050_Pow_x_n/CSharp/1_Binary_Exponentiation.cs
public class Solution {
    private double BinaryExp(double x, long n) {
        // Base case, to stop recursive calls.
        if (n == 0) {
            return 1;
        }

        // Handle case where, n < 0.
        if (n < 0) {
            return 1.0 / BinaryExp(x, -1 * n);
        }

        // Perform Binary Exponentiation.
        // If 'n' is odd we perform Binary Exponentiation on 'n - 1' and
        // multiply result with 'x'.
        if (n % 2 == 1) {
            return x * BinaryExp(x * x, (n - 1) / 2);
        }
        // Otherwise we calculate result by performing Binary Exponentiation on
        // 'n'.
        else {
            return BinaryExp(x * x, n / 2);
        }
    }

    public double MyPow(double x, int n) {
        // Call recursive function with correct types.
        return BinaryExp(x, (long)n);
    }
}
